llvm/test/lab3/bodrov_daniil/test.mir

```diff
@@ -0,0 +1,878 @@
+  - { reg: '$edx', virtual-reg: '' }
+frameInfo:
+  isFrameAddressTaken: false
+  isReturnAddressTaken: false
+  hasStackMap:     false
+  hasPatchPoint:   false
+  stackSize:       0
+  offsetAdjustment: 0
+  maxAlignment:    1
+  adjustsStack:    false
+  hasCalls:        false
+  stackProtector:  ''
+  functionContext: ''
+  maxCallFrameSize: 0
+  cvBytesOfCalleeSavedRegisters: 0
+  hasOpaqueSPAdjustment: false
+  hasVAStart:      false
+  hasMustTailInVarArgFunc: false
+  hasTailCall:     false
+  localFrameSize:  0
+  savePoint:       ''
+  restorePoint:    ''
+fixedStack:      []
+stack:           []
+entry_values:    []
+callSites:       []
+debugValueSubstitutions: []
+constants:       []
+machineFunctionInfo: {}
+body:             |
+  bb.0.entry:
+    successors: %bb.2(0x50000000), %bb.1(0x30000000)
+    liveins: $edi, $edx, $esi
+  
+    MOV64mi32 $rip, 1, $noreg, @ic, $noreg, 0 :: (store (s64) into @ic, !tbaa !5)
+    TEST32rr renamable $edi, renamable $edi, implicit-def $eflags
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 3, implicit-def $eflags
+    JCC_1 %bb.1, 14, implicit $eflags
+  
+  bb.2.for.body.preheader:
+    successors: %bb.3(0x40000000), %bb.7(0x40000000)
+    liveins: $edi, $edx, $esi
+  
+    $ecx = MOV32rr $edi
+    renamable $ecx = AND32ri killed renamable $ecx, 3, implicit-def dead $eflags
+    CMP32ri renamable $edi, 4, implicit-def $eflags
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 4, implicit-def $eflags
+    JCC_1 %bb.7, 3, implicit $eflags
+  
+  bb.3:
+    successors: %bb.4(0x80000000)
+    liveins: $ecx, $edx, $esi
+  
+    renamable $eax = XOR32rr undef $eax, undef $eax, implicit-def dead $eflags
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 1, implicit-def $eflags
+  
+  bb.4.for.cond.cleanup.loopexit.unr-lcssa:
+    successors: %bb.6(0x30000000), %bb.5(0x50000000)
+    liveins: $eax, $ecx, $edx, $esi
+  
+    TEST32rr renamable $ecx, renamable $ecx, implicit-def $eflags
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 2, implicit-def $eflags
+    JCC_1 %bb.6, 4, implicit $eflags
+  
+  bb.5.for.body.epil (align 16):
+    successors: %bb.6(0x04000000), %bb.5(0x7c000000)
+    liveins: $eax, $ecx, $edx, $esi
+  
+    CMP32rr renamable $eax, renamable $esi, implicit-def $eflags
+    renamable $edi = MOV32ri 0
+    renamable $edi = CMOV32rr killed renamable $edi, renamable $edx, 12, implicit $eflags
+    renamable $eax = nsw ADD32rr killed renamable $eax, killed renamable $edi, implicit-def dead $eflags
+    renamable $ecx = DEC32r killed renamable $ecx, implicit-def $eflags
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 6, implicit-def $eflags
+    JCC_1 %bb.5, 5, implicit $eflags
+  
+  bb.6.for.cond.cleanup:
+    liveins: $eax
+  
+    ; CHECK: MOV64mr %0, &ic
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 1, implicit-def $eflags
+    RET64 $eax
+  
+  bb.1:
+    ; CHECK: MOV64mr %0, &ic
+    renamable $eax = XOR32rr undef $eax, undef $eax, implicit-def dead $eflags
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 2, implicit-def $eflags
```

Comment:
So, you first move counter value from register into memory and after it increment the counter again?